From dd1da4f525277177a39abd5c70467c6368de9d57 Mon Sep 17 00:00:00 2001
From: Brad Ho <brad_ho@phoenix.com>
Date: Fri, 18 Feb 2022 16:50:57 +0800
Subject: [PATCH] Merge from server-gold

Note:
 1. Change organization to phoenix
 2. Support TLSv1.3
 3. Remove non support SSL protocol
---
 include/ssl_key_handler.hpp | 24 +++++++++++-------------
 1 file changed, 11 insertions(+), 13 deletions(-)

diff --git a/include/ssl_key_handler.hpp b/include/ssl_key_handler.hpp
index 7710ae2..6a9bd98 100644
--- a/include/ssl_key_handler.hpp
+++ b/include/ssl_key_handler.hpp
@@ -264,7 +264,7 @@ inline void generateSslCertificate(const std::string& filepath,
                 reinterpret_cast<const unsigned char*>("US"), -1, -1, 0);
             X509_NAME_add_entry_by_txt(
                 name, "O", MBSTRING_ASC,
-                reinterpret_cast<const unsigned char*>("OpenBMC"), -1, -1, 0);
+                reinterpret_cast<const unsigned char*>("Phoenix Technologies Ltd."), -1, -1, 0);
             X509_NAME_add_entry_by_txt(
                 name, "CN", MBSTRING_ASC,
                 reinterpret_cast<const unsigned char*>(cn.c_str()), -1, -1, 0);
@@ -371,7 +371,7 @@ inline std::shared_ptr<boost::asio::ssl::context>
     // unauthenticated endpoints
     // mSslContext->set_verify_mode(boost::asio::ssl::verify_peer);
 
-    SSL_CTX_set_options(mSslContext->native_handle(), SSL_OP_NO_RENEGOTIATION);
+    SSL_CTX_set_options(mSslContext->native_handle(), SSL_OP_CIPHER_SERVER_PREFERENCE);
 
     BMCWEB_LOG_DEBUG << "Using default TrustStore location: " << trustStorePath;
     mSslContext->add_verify_path(trustStorePath);
@@ -385,27 +385,25 @@ inline std::shared_ptr<boost::asio::ssl::context>
     // There is a pull request to add this.  Once this is included in an asio
     // drop, use the right way
     // http://stackoverflow.com/questions/18929049/boost-asio-with-ecdsa-certificate-issue
-    if (SSL_CTX_set_ecdh_auto(mSslContext->native_handle(), 1) != 1)
+    if (1 != SSL_CTX_set1_curves_list(mSslContext->native_handle(), "X448:X25519:secp521r1:secp384r1:prime256v1"))
     {
-        BMCWEB_LOG_ERROR << "Error setting tmp ecdh list\n";
+        BMCWEB_LOG_ERROR << "Error 2 setting tmp ecdh list\n";
     }
-
     std::string mozillaModern = "ECDHE-ECDSA-AES256-GCM-SHA384:"
                                 "ECDHE-RSA-AES256-GCM-SHA384:"
-                                "ECDHE-ECDSA-CHACHA20-POLY1305:"
-                                "ECDHE-RSA-CHACHA20-POLY1305:"
                                 "ECDHE-ECDSA-AES128-GCM-SHA256:"
-                                "ECDHE-RSA-AES128-GCM-SHA256:"
-                                "ECDHE-ECDSA-AES256-SHA384:"
-                                "ECDHE-RSA-AES256-SHA384:"
-                                "ECDHE-ECDSA-AES128-SHA256:"
-                                "ECDHE-RSA-AES128-SHA256";
-
+                                "ECDHE-RSA-AES128-GCM-SHA256";
     if (SSL_CTX_set_cipher_list(mSslContext->native_handle(),
                                 mozillaModern.c_str()) != 1)
     {
         BMCWEB_LOG_ERROR << "Error setting cipher list\n";
     }
+
+    if (SSL_CTX_set_ciphersuites(mSslContext->native_handle(),
+                "TLS_AES_256_GCM_SHA384:TLS_AES_128_GCM_SHA256") != 1)
+    {
+        BMCWEB_LOG_ERROR << "Error setting TLSv1.3 ciphersuites list\n";
+    }
     return mSslContext;
 }
 } // namespace ensuressl
